### ISSUE TITLE
Async not working Rails 7 with puma CLI

### ISSUE THREAD

**Issue by jcrate** [bensheldon/good_job#685]
I have a project using good_job 2.15 on Rails 6.1 using async processing with Puma server (5.6.1) and it's working well. I attempted the same setup in a Rails 7 project with good_job 3.3.2 and puma 5.6.4 and good_job never activates.  I'm using the same postgresql installation for both (development setup). 

application.rb
```
    config.active_job.queue_adapter = :good_job
    config.good_job = { execution_mode: :async, max_threads: 4, poll_interval: 30 }
```

puma.rb
```
before_fork do
  GoodJob.shutdown
end

on_worker_boot do
  GoodJob.restart
end

on_worker_shutdown do
  GoodJob.shutdown
end
```
I'm not even sure that's necessary, neither system is using puma workers, only threads.

When puma is started on the Rails 6 app, the following appears in the development log.
```
[GoodJob] [33083] GoodJob started scheduler with queues=* max_threads=4.
   (0.9ms)  SET application_name = 'GoodJob::Notifier'
   (2.6ms)  LISTEN good_job
[GoodJob] Notifier subscribed with LISTEN
```

The rails 7 database doesn't show anything in the log when Puma is started, but if I use `bundle exec good_job start` I get this:
```
[GoodJob] [33139] GoodJob started scheduler with queues=* max_threads=4.
   (0.3ms)  SET application_name = 'GoodJob::Notifier'
   (0.3ms)  LISTEN good_job
[GoodJob] Notifier subscribed with LISTEN
```

GoodJob does process jobs once running, but it would be nice if the async mode worked.

**Comment by bensheldon** [comment]
@jcrate thanks for opening this? Any chance you could try upgrading just Rails or just GoodJob to help isolate down which change caused the change? 

That would help me because auto loading can be tricky and app-specific. Thanks!

**Comment by jcrate** [comment]
I took the working Rails 6.1 / GoodJob 2.15.1 app and tested with combinations of Rails 6/7, GoodJob 2.15/3.3, and Ruby 2.7.2 and 3.0.2.  GoodJob started up in async mode with every combination. 

I created a new rails 7 app with a new PostgreSQL db, installed good_job, ran the migration and copied the good_job async config to application.rb. Launching Puma does not start good_job. So I created a new rails 6.1.6 app and did the same thing, and again, good_job does not start in async mode when launching puma. 

I decided to try puma with workers to see what happened, and got this error:
```
[76432] WARNING hook before_fork failed with exception (NameError) uninitialized constant #<Class:#<Puma::DSL:0x00007f88ea9b62f0>>::GoodJob
[76433] WARNING hook before_worker_boot failed with exception (NameError) uninitialized constant #<Class:#<Puma::DSL:0x00007f88ea9b62f0>>::GoodJob
```

If this is happening with the thread-mode auto loading I'm not seeing any messages.

**Comment by bensheldon** [comment]
@jcrate Let me ask: When you run `bin/rails s`, are you then loading `http://localhost:3000` in your browser? Making the http request is what triggers GoodJob async starting for me when I set up a new app to try to reproduce (Ruby 3.1.2, Rails 6.1.6.1 and 7.0.3.1, GoodJob 3.3.2).

My reproduction commands are:

```shell
rails new demo_685 -d postgresql
bundle add good_job
bin/rails g good_job:install
bin/setup
# Add to application.rb: config.active_job.queue_adapter = :good_job

bin/rails s
```

If making the initial HTTP Request is what starts it up, then I think we just have a documentation improvement. Rails' development mode is getting better and better and deferring initialization/autoloading for faster boot times.

(I'm ignoring those uninitialized constant errors for the moment, those are weird and I'm not seeing them at all)

**Comment by bensheldon** [comment]
> Maybe I'm the only one left using `bundle exec puma` to launch my apps. 

You're definitely the first person to open an issue on GoodJob about it 😄  That was the detail I needed to reproduce it 🙌🏻 

Yep, GoodJob doesn't detect async when run directly from `$ puma`. I have a PR to fix that: #686

**Comment by bensheldon** [comment]
Released in [v3.3.3](https://github.com/bensheldon/good_job/releases/tag/v3.3.3).

@jcrate could you verify that releases addresses the problem for you? Thank you!

**Comment by jcrate** [comment]
Yes, v3.3.3 now starts GoodJob in async mode when the rails app is served with `bundle exec puma`, same as with `rails s`.

Thanks!
